### ISSUE TITLE
Randomize Design

### ISSUE THREAD

**Issue by ghost** [buddhi1980/mandelbulber2#278]
We love the randomize color palette option.
How about another option for the fractal solid voxels.
Please implement a random button that selects from the different formulas and transforms and populates parameter data.  
Each click of the new "Randomize Design" button uses pseudo-random number generators to assure creation of unique content.

**Comment by buddhi1980** [comment]
Have you seen Mutagen in JWildfire program?

**Comment by mancoast** [comment]
I havevt seen it, but I like the name.
Maybe we can call it 'transmutate' button

**Comment by mclarekin** [comment]
Apohysis also  has a mutagen type thing.

I would suggest that Mandelbulber is much more complex for the simplified
approach used in other programs
 The  working range of a  specific parameter is sometimes very dependent on
other parameter settings.

There :

- may have to be a selection of formula related mutagens, constructed from
a working knowledge of the formula specific parameters.

- may require both a Basic and an Advanced option.

I consider it to be low priority at the moment.

**Comment by mancoast** [comment]
Reference mandelbulb 3D source code:
https://github.com/thargor6/mb3d/blob/master/mutagen/MutaGen.pas

This enables beginners to immediately dive in and make original content.

**Comment by mancoast** [comment]
@buddhi1980 Can we 'pseudo' implement this feature by simply loading a random example, then randomizing the color palette, and randomizing the camera target?

**Comment by zebastian** [comment]
my five cents on this topic:

its hard / impossible to make educated guesses about a fractal formula parameter and its effect in the first place. A fine tuned Trial and error machine could help:
 
A randomizer could be implemented the following way
iterate till 'cool':
  1. take a random parameter and change by random amount
  2. apply heuristics and check if result is better than before (than keep changes) else discard

the heuristics can be managable by the user and may include
- render time
- amount bad DE
- color diversity (areas of same color and areas of changing color)
- distance diversity (close points and far away points)
- ...
there should be reasonable ways to measure these heuristics and the user should be able to weigh each heuristic.

with this hill climbing (https://en.wikipedia.org/wiki/Hill_climbing) algorithm it should be possible to find interesting absolutely new settings. This way all possible parameters are taken into account (though there should be a blacklist of parameters which should not be taken into account like stereoscopic rendering). 
Should be moderate programming effort. How well this will work i dont know.
What do you think?

**Comment by mclarekin** [comment]
a) I think we would need to make several randomizers. Basic (controlled
limited  randomization) to Complex (suitable for an experienced user.)

. basic is UIs that will cover the situations that are most likely to
produce good results. Maybe a few analytical linear and log ones. OpenCL
implementation makes this feasible .

 The heuristics control with bad DE may need to take into account the
proportion of fractal surface in the image.

b) There are so many different methods  ( to find good settings I use to
test the combinations  by making morphing animations  )

We can use  groupboxes or populate a table like we do for animation table
i.e. add_parameter  to the Randomizer  UI table,

The user can randomize parameter between either,  a  coded default max &
min,  or user defined.

For a basic UI the choice of transforms would be restricted to ones that
will work well together.

- the most basic is to do a  randomization on a single formula's
parameters. This could be done by adding an enable group box , with
parameters set up to randomize between either,  coded default max & min,
or user defined. Enable/disable individual parameters to randomize.

- another basic method is a controlled randomization of  formulas and
transforms in the slots. This method includes randomizing  stop, start,
number of repeat iterations, and weight.

- then we can make much more complex methods, involving multiple  slots and
transforms. together with a complex table(,   that can also be used as a
mega hybrid table.)

I would suggest trialling some basic methods once we have openCL speed.



c) Apophysis  had a 2 way approach.

First you could load a random batch ie  50.  These were displayed as a
column of thumbnails. You would look at them at see that maybe only 1
looked interesting. The batch  was all randomly constructed from a user
defined preference list of transforms. (An experienced used could get a
better success rate,)

Secondly, you could then  mutate the chosen random image. The original
image was in the UI center , and surrounded by 8 different mutated images.
You could then tweak a parameter and see the  way that parameter interacted
with the mutations. There were quite a lot of user controls for the two
stage process, however Apophysis is much simpler than Mandelbulber.

**Comment by taurus66** [comment]
I use random fields for several values. They are currently generated in a simple spread sheet. They depend on the initial value and the amount of variation. So in this case an additional input slot would be needed for the wanted variation - maybe global within the measurement module. 
With the following simple calculation, I make sure, to get a randomized value (or vector), that has a range of initial value +- variation: p--initial value; v--variation; r--random number between 0 and 1
p(new)=p-v+2*v*r
Easy to see that at r 0-0.5 -> new value below initial and r 0.5-1 -> new value above initial.
As I was pretty happy with Mandelbulber v1, I don't need any slider to input variation. A float with enough digits would do much better for me. I hope it is not a problem, that some values (like rotations or constant multiplier) only accept values, with a limited number of digits.
I also do not need a full featured module like the mutagen in m3d 1.91. I need full control and not that slider ease of use - not a popular point of view, but my humble opinion.

**Comment by mancoast** [comment]
The ability to randomize that enables everyone to instantly create.

**Comment by TasMania17** [comment]
I found an easy way to achieve this. I used the Unix 'rand' utility to choose say 3 fractals from a possible 98. (It's over 100 now). I made a list of all the primary fractals (not Formulas 4D, transforms 4D, Transforms 4D or Hybrid colouring).
Then I ran the command:
rand -N 3 -M 98 -e -d '\n' 
and looked up the 3 fractals from my list.
I did produce fairly good results.
There are a possible 152,000 combinations of 3 from 98.
I'm sure there are equivalent Windows utilities to rand.
https://fractalforums.org/mandelbulber/14/experiment-using-a-random-number-generator-to-choose-fractals/2501
